platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/ConfigurationCacheGracefulDegradationIntegrationTest.groovy

```diff
@@ -0,0 +1,244 @@
+                        it.taskDependencies
+                        println "Task's dependencies accessed!"
+                    }
+                }
+            }
+        """
+
+        when:
+        configurationCacheRun("a", *args)
+
+        then:
+        configurationCache.assertNoConfigurationCache()
+        problems.assertResultHtmlReportHasProblems(result) {
+            totalProblemsCount = expectedProblems.size()
+            expectedProblems.forEach { withProblem(it) }
+            withIncompatibleTask(":a", degradationReason)
+        }
+
+        where:
+        args                                                          | expectedOutputs                                               | expectedProblems                                                                                                | degradationReason
+        ["-DaccessTaskProject=true"]                                  | ["Task's project accessed!"]                                  | ["Build file 'build.gradle': line 10: invocation of 'Task.project' at execution time is unsupported."]          | "Project access."
+        ["-DaccessTaskProject=true", "-DaccessTaskDependencies=true"] | ["Task's project accessed!", "Task's dependencies accessed!"] | ["Build file 'build.gradle': line 10: invocation of 'Task.project' at execution time is unsupported.",
+                                                                                                                                         "Build file 'build.gradle': line 14: invocation of 'Task.taskDependencies' at execution time is unsupported."] | "Project access, TaskDependencies access."
+    }
+
+    def "CC problems in incompatible tasks are not hidden by CC degradation"() {
+        def configurationCache = newConfigurationCacheFixture()
+        buildFile """
+            ${generateDegradationController()}
+            tasks.register("foo") { task ->
+                degradation.requireConfigurationCacheDegradation(task, provider { "Project access" })
+                doLast {
+                    println "Hello from foo \${project.path}"
+                }
+            }
+
+            tasks.register("bar") {
+                doLast {
+                    println "Hello from bar \${project.path}"
+                }
+            }
+        """
+
+        when:
+        configurationCacheFails "foo", "bar"
+
+        then:
+        configurationCache.assertNoConfigurationCache()
+        problems.assertFailureHtmlReportHasProblems(failure) {
+            totalProblemsCount = 2
+            withProblem("Build file 'build.gradle': line 11: invocation of 'Task.project' at execution time is unsupported.")
+            withProblem("Build file 'build.gradle': line 5: invocation of 'Task.project' at execution time is unsupported.")
+            withIncompatibleTask(":foo", "Project access.")
+        }
+    }
+
+    def "a task in included build can require CC degradation"() {
```

Comment:
🧪 I haven't seen a test for "a task in build logic can require CC degradation but it won't affect main build"

Reply:
I added such test here: https://github.com/gradle/gradle/pull/33444/commits/6d55d755bb3e0557df0924e9cf53dd286b45e4ac